src/libsyntax_ext/deriving/clone.rs

```diff
@@ -24,13 +25,54 @@ pub fn expand_deriving_clone(cx: &mut ExtCtxt,
                              item: &Annotatable,
                              push: &mut FnMut(Annotatable))
 {
+    // check if we can use a short form
+    //
+    // the short form is `fn clone(&self) -> Self { *self }`
+    //
+    // we can use the short form if:
+    // - the item is Copy (unfortunately, all we can check is whether it's also deriving Copy)
+    // - there are no generic parameters (after specialization this limitation can be removed)
+    //      if we used the short form with generics, we'd have to bound the generics with
+    //      Clone + Copy, and then there'd be no Clone impl at all if the user fills in something
+    //      that is Clone but not Copy. and until specialization we can't write both impls.
+    let bounds;
+    let substructure;
+    let nested_match;
+    match *item {
+        Annotatable::Item(ref item) => {
+            match item.node {
+                ast::ItemStruct(_, ast::Generics { ref ty_params, .. })
+                    | ast::ItemEnum(_, ast::Generics { ref ty_params, .. })
+                    if ty_params.is_empty()
+                        && (|| item.attrs.iter().any(|a| a.name() == "derive_Copy"))() => {
+
+                    bounds = vec![Literal(path_std!(cx, core::marker::Copy))];
+                    substructure = combine_substructure(Box::new(|c, s, _| {
+                        cs_shallow_clone(c, s)
+                    }));
+                    nested_match = false;
+                }
+
+                _ => {
+                    bounds = vec![];
+                    substructure = combine_substructure(Box::new(|c, s, sub| {
+                        cs_deep_clone("Clone", c, s, sub)
+                    }));
+                    nested_match = true;
+                }
+            }
+        }
+
+        _ => cx.span_bug(span, "#[derive(Clone)] on trait item or impl item")
```

Comment:
> I'm wrong again because it could be a different kind of Item. So it should just ICE with "... on non-struct/enum item"

I think you should just have something like

```
match *item {
    Annotatable::Item(ast::Item { node: ast::ItemStruct { ... } }) |
    Annotatable::Item(ast::Item { node: ast::ItemEnum { ... } })
    if /* the other conditions */ => {
         // this is copy branch
    }
    _ => /* the clone branch */
}
```